benchmarks/bm_runner.py

```diff
@@ -0,0 +1,371 @@
+        if not publish_dir.is_dir():
+            message = (
+                f"Input 'publish directory' is not a directory: {publish_dir}"
+            )
+            raise NotADirectoryError(message)
+        publish_subdir = (
+            publish_dir
+            / f"{run_type}_{datetime.now().strftime('%Y%m%d_%H%M%S')}"
+        )
+        publish_subdir.mkdir()
+
+        # Activate on demand benchmarks (C/SPerf are deactivated for
+        #  'standard' runs).
+        environ["ON_DEMAND_BENCHMARKS"] = "True"
+        commit_range = "upstream/main^!"
+
+        asv_command = (
+            ASV_HARNESS.format(posargs=commit_range) + f" --bench={run_type}"
+        )
+        # C/SPerf benchmarks are much bigger than the CI ones:
+        # Don't fail the whole run if memory blows on 1 benchmark.
+        asv_command = asv_command.replace(" --strict", "")
+        # Only do a single round.
+        asv_command = re.sub(r"rounds=\d", "rounds=1", asv_command)
+        subprocess.run([*asv_command.split(" "), *args.asv_args])
+
+        asv_command = f"asv publish {commit_range} --html-dir={publish_subdir}"
+        subprocess.run(asv_command.split(" "))
+
+        # Print completion message.
+        location = Path().cwd() / ".asv"
+        print(
+            f'New ASV results for "{run_type}".\n'
+            f'See "{publish_subdir}",'
+            f'\n  or JSON files under "{location / "results"}".'
+        )
+
+
+class CPerf(_CSPerf):
+    name = "cperf"
+    description = _CSPerf.description.format("CPerf")
+
+    @staticmethod
+    def func(args: argparse.Namespace) -> None:
+        _CSPerf.csperf(args, "cperf")
+
+
+class SPerf(_CSPerf):
+    name = "sperf"
+    description = _CSPerf.description.format("SPerf")
+
+    @staticmethod
+    def func(args: argparse.Namespace) -> None:
+        _CSPerf.csperf(args, "sperf")
+
+
+class Custom(_SubParserGenerator):
+    name = "custom"
+    description = (
+        "Run ASV with the input **ASV sub-command**, without any preset "
+        "arguments - must all be supplied by the user. So just like running "
+        "ASV manually, with the convenience of re-using the runner's "
+        "scripted setup steps."
+    )
+
+    def add_arguments(self) -> None:
+        self.subparser.add_argument(
+            "asv_sub_command",
+            type=str,
+            help="The ASV command to run.",
+        )
+
+    @staticmethod
+    def func(args: argparse.Namespace) -> None:
+        _setup_common()
+        subprocess.run(["asv", args.asv_sub_command, *args.asv_args])
+
+
+def main():
+    parser = ArgumentParser(
+        description="Run the Iris performance benchmarks (using Airspeed Velocity).",
+        epilog="More help is available within each sub-command.",
+    )
+    subparsers = parser.add_subparsers()
+
+    for gen in (Overnight, Branch, CPerf, SPerf, Custom):
+        _ = gen(subparsers).subparser
+
+    parsed = parser.parse_args()
+    parsed.func(parsed)
+    pass
```

Comment:
Is this necessary? I assume it's not doing anything. You don't need a placeholder for future code as you already have code here

Reply:
Thanks for spotting. Left over from when I wanted a breakpoint there.

f469f0f705c4875d924b25ab9b79b6d0a53872d4